src/hotspot/share/gc/shared/workerThread.cpp

```diff
@@ -56,17 +66,34 @@ void WorkerTaskDispatcher::coordinator_distribute_task(WorkerTask* task, uint nu
   _started = 0;
 }
 
+void WorkerTaskDispatcher::caller_run_task() {
+  // Run at least one task in the caller.
+  // Then see if we have other tasks that caller can run.
+  do {
+    internal_run_task(false);
+  } while (_start_semaphore.trywait());
+}
+
 void WorkerTaskDispatcher::worker_run_task() {
   // Wait for the coordinator to dispatch a task.
   _start_semaphore.wait();
+  internal_run_task(true);
+}
 
+void WorkerTaskDispatcher::internal_run_task(bool is_worker) {
   // Get and set worker id.
   const uint worker_id = Atomic::fetch_then_add(&_started, 1u);
-  WorkerThread::set_worker_id(worker_id);
+  if (is_worker) {
+    WorkerThread::set_worker_id(worker_id);
+  }
 
   // Run task.
-  GCIdMark gc_id_mark(_task->gc_id());
-  _task->work(worker_id);
+  if (is_worker || Thread::current()->is_Named_thread()) {
+    GCIdMark gc_id_mark(_task->gc_id());
+    _task->work(worker_id);
+  } else {
+    _task->work(worker_id);
```

Comment:
If we don't set the GCId, the logs may be obscure. But I don't know whether such non named thread exists.